### PR TITLE
pkg/lwip: fix netdev concurrency issues

### PR DESCRIPTION
**Pull request by jia200x**
### Contribution description

This PR fixes a concurrency issue in LWIP, as a result of calling `netdev` functions from different context.
Traditionally our netdev driver model assumes implicitly that all `netdev` functions run on the same thread. As it is now, LWIP calls all send functions from the caller thread (e.g `main`), but processes ISR on a higher priority thread. This not only creates concurrency issues in existing devices, but renders other modules such as the SubMAC unusable...

To (temporally) solve this, I implemented some recursive mutex lock functions to acquire the device before calling any internal operation. In the future it would be better to schedule all transmissions from a single thread.

Although this won't make the current situation worse, I'm assuming LWIP uses mutex internally to prevent concurrency issues between LWIP functions called from different context. If this doesn't hold, we will need to proceed with the single thread solution.

EDIT: I also adapted the `netdev` return calls to the newer API. Returning 0 is not considered an error anymore (which is what `netdev_ieee802154_submac` does).

### Testing procedure
Test that LWIP still works as expected (e.g `tests/lwip`).

### Issues/PRs references
Found while porting the KW2XRF to LWIP #18465

### CODE REVIEW

**Comment by jia200x** [comment]
> Although this won't make the current situation worse, I'm assuming LWIP uses mutex internally to prevent concurrency issues between LWIP functions called from different context. If this doesn't hold, we will need to proceed with the single thread solution.

@miri64 does this assumption holds?

**Comment by maribu** [comment]
lwIP internally synchronizes all acceses via a global lock and all internal functions assert that the lock is taken when they are called. I *guess* that lock is implemented via a mutex in RIOT, but I haven dived deep enough into the implementation to say be sure.

File: pkg/lwip/contrib/netdev/lwip_netdev.c [hunk 1]
```diff
@@ -65,6 +65,7 @@ static err_t _ieee802154_link_output(struct netif *netif, struct pbuf *p);
 static void _event_cb(netdev_t *dev, netdev_event_t event);
 static void *_event_loop(void *arg);
 
+
```
Comment (maribu) [inline pkg/lwip/contrib/netdev/lwip_netdev.c:68]:
```suggestion
```

There already is an empty above

**Review by maribu — APPROVED** [review]
ACK. Code looks good and I trust your testing.

I believe this can fix a lot of sporadic and hard to reproduce issues.

**Comment by chrysn** [comment]
This has been requested for backporting. I don't see the big picture of its complexity yet, but am marking it for "needs backport" if only to make sure I'll have it in the "known issues" section.

**Comment by jia200x** [comment]
amended and squashed direclty

**Pull request merged**